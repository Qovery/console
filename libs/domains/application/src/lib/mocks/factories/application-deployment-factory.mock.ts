import { DeploymentHistoryApplication, StateEnum } from 'qovery-typescript-axios'
import { Chance } from 'chance'

const chance = new Chance()

export const applicationDeploymentsFactoryMock = (howMany: number): DeploymentHistoryApplication[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: index.toString(),
    created_at: new Date().toString(),
    updated_at: new Date().toString(),
    name: chance.name(),
    status: chance.pickone(Object.values([StateEnum.DEPLOYED, StateEnum.RUNNING, StateEnum.STOP_ERROR])),
  }))
