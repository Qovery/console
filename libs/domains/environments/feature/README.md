# domains-environments-feature

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test domains-environments-feature` to execute the unit tests via [Jest](https://jestjs.io).
