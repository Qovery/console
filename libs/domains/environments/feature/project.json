{
  "name": "domains-environments-feature",
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/domains/environments/feature/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "test": {
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "coverage": true
        }
      }
    }
  }
}
