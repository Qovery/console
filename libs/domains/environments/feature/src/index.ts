export * from './lib/environment-deployment-status-label/environment-deployment-status-label'
export * from './lib/environment-mode/environment-mode'
export * from './lib/environment-state-chip/environment-state-chip'
export * from './lib/environment-list/environment-list'
export * from './lib/environment-avatar/environment-avatar'
export * from './lib/terraform-export-modal/terraform-export-modal'
export * from './lib/create-clone-environment-modal/create-clone-environment-modal'
export * from './lib/environment-action-toolbar/environment-action-toolbar'
export * from './lib/hooks/use-deployment-status/use-deployment-status'
export * from './lib/hooks/use-deployment-stages/use-deployment-stages'
export * from './lib/hooks/use-environment/use-environment'
export * from './lib/hooks/use-edit-environment/use-edit-environment'
export * from './lib/hooks/use-delete-environment/use-delete-environment'
export * from './lib/hooks/use-list-statuses/use-list-statuses'
export * from './lib/hooks/use-environments/use-environments'
export * from './lib/hooks/use-create-environment/use-create-environment'
export * from './lib/hooks/use-clone-environment/use-clone-environment'
export * from './lib/hooks/use-create-deployment-stage/use-create-deployment-stage'
export * from './lib/hooks/use-edit-deployment-stage/use-edit-deployment-stage'
export * from './lib/hooks/use-delete-deployment-stage/use-delete-deployment-stage'
export * from './lib/hooks/use-list-deployment-stages/use-list-deployment-stages'
export * from './lib/hooks/use-attach-service-to-deployment-stage/use-attach-service-to-deployment-stage'
export * from './lib/hooks/use-move-deployment-stage/use-move-deployment-stage'
export * from './lib/hooks/use-deployment-history-legacy/use-deployment-history-legacy'
export * from './lib/hooks/use-deployment-history/use-deployment-history'
export * from './lib/hooks/use-deployment-history-execution-id/use-deployment-history-execution-id'
export * from './lib/hooks/use-list-database-configurations/use-list-database-configurations'
export * from './lib/hooks/use-deployment-rule/use-deployment-rule'
export * from './lib/hooks/use-edit-deployment-rule/use-edit-deployment-rule'
export * from './lib/hooks/use-cancel-deployment-environment/use-cancel-deployment-environment'
export * from './lib/hooks/use-deploy-environment/use-deploy-environment'
export * from './lib/hooks/use-check-dockerfile/use-check-dockerfile'
export * from './lib/hooks/use-lifecycle-templates/use-lifecycle-templates'
export * from './lib/hooks/use-lifecycle-template/use-lifecycle-template'
export * from './lib/hooks/use-deploy-all-services/use-deploy-all-services'
