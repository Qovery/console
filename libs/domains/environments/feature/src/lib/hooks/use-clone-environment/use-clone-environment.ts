import { useMutation, useQueryClient } from '@tanstack/react-query'
import { mutations } from '@qovery/domains/environments/data-access'
import { queries } from '@qovery/state/util-queries'

export function useCloneEnvironment() {
  const queryClient = useQueryClient()

  return useMutation(mutations.cloneEnvironment, {
    onSuccess({ project }) {
      queryClient.invalidateQueries({
        queryKey: queries.environments.list({ projectId: project.id }).queryKey,
      })
    },
    meta: {
      notifyOnSuccess: {
        title: 'Your environment has been successfully cloned',
      },
      notifyOnError: true,
    },
  })
}

export default useCloneEnvironment
