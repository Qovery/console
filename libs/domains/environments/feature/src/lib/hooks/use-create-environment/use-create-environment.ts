import { useMutation, useQueryClient } from '@tanstack/react-query'
import { mutations } from '@qovery/domains/environments/data-access'
import { queries } from '@qovery/state/util-queries'

export function useCreateEnvironment() {
  const queryClient = useQueryClient()

  return useMutation(mutations.createEnvironment, {
    onSuccess(_, { projectId }) {
      queryClient.invalidateQueries({
        queryKey: queries.environments.list({ projectId }).queryKey,
      })
    },
    meta: {
      notifyOnSuccess: {
        title: 'Your environment has been successfully created',
      },
      notifyOnError: true,
    },
  })
}

export default useCreateEnvironment
