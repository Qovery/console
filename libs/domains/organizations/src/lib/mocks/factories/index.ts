export * from './organizations-factory.mock'
