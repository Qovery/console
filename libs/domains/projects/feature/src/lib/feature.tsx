/* eslint-disable-next-line */
export interface FeatureProps {}

export function Feature(props: FeatureProps) {
  return (
    <div>
      <h1>Welcome to Feature!</h1>
    </div>
  )
}

export default Feature
