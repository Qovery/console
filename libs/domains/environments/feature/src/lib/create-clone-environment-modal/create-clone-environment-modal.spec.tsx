import { EnvironmentModeEnum } from 'qovery-typescript-axios'
import selectEvent from 'react-select-event'
import * as clustersDomain from '@qovery/domains/clusters/feature'
import { clusterFactoryMock, environmentFactoryMock } from '@qovery/shared/factories'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import * as useCloneEnvironment from '../hooks/use-clone-environment/use-clone-environment'
import * as useCreateEnvironment from '../hooks/use-create-environment/use-create-environment'
import CreateCloneEnvironmentModal, { type CreateCloneEnvironmentModalProps } from './create-clone-environment-modal'

let props: CreateCloneEnvironmentModalProps

const useClustersMockSpy = jest.spyOn(clustersDomain, 'useClusters') as jest.Mock
const useCloneEnvironmentMockSpy = jest.spyOn(useCloneEnvironment, 'useCloneEnvironment') as jest.Mock
const useCreateEnvironmentMockSpy = jest.spyOn(useCreateEnvironment, 'useCreateEnvironment') as jest.Mock

const mockClusters = clusterFactoryMock(3, 'AWS')

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ projectId: '1', organizationId: '0' }),
}))

describe('CreateCloneEnvironmentModal', () => {
  beforeEach(() => {
    props = {
      onClose: jest.fn(),
      environmentToClone: undefined,
      organizationId: '0',
      projectId: '1',
    }

    useClustersMockSpy.mockReturnValue({
      data: mockClusters,
    })
    useCloneEnvironmentMockSpy.mockReturnValue({
      mutateAsync: jest.fn(() => Promise.resolve({ id: '1' })),
    })
    useCreateEnvironmentMockSpy.mockReturnValue({
      mutateAsync: jest.fn(() => Promise.resolve({ id: '1' })),
    })
  })

  it('should render successfully', () => {
    const { baseElement } = renderWithProviders(<CreateCloneEnvironmentModal {...props} />)
    expect(baseElement).toBeTruthy()
  })

  it('should match snapshots', () => {
    const { baseElement } = renderWithProviders(<CreateCloneEnvironmentModal {...props} />)
    expect(baseElement).toMatchSnapshot()
  })

  describe('creation mode', function () {
    it('should submit form on click on button', async () => {
      const { userEvent } = renderWithProviders(<CreateCloneEnvironmentModal {...props} />)

      const input = screen.getByTestId('input-text')

      await userEvent.type(input, 'test')

      await selectEvent.select(screen.getByLabelText('Cluster'), `${mockClusters[2].name} - Managed (EKS)`, {
        container: document.body,
      })

      await selectEvent.select(screen.getByLabelText('Type'), 'Staging', { container: document.body })

      const submitButton = screen.getByTestId('submit-button')
      await userEvent.click(submitButton)

      expect(useCreateEnvironmentMockSpy().mutateAsync).toHaveBeenCalledWith({
        projectId: '1',
        payload: {
          cluster: mockClusters[2].id,
          mode: EnvironmentModeEnum.STAGING,
          name: 'test',
        },
      })
    })
  })

  describe('cloning mode', function () {
    it('should submit form on click on button', async () => {
      const mockEnv = environmentFactoryMock(1)[0]
      const { userEvent } = renderWithProviders(<CreateCloneEnvironmentModal {...props} environmentToClone={mockEnv} />)

      const inputs = screen.getAllByTestId('input-text')

      await userEvent.clear(inputs[1])
      await userEvent.type(inputs[1], 'test')

      await selectEvent.select(screen.getByLabelText('Cluster'), `${mockClusters[2].name} - Managed (EKS)`, {
        container: document.body,
      })

      await selectEvent.select(screen.getByLabelText('Type'), 'Staging', { container: document.body })

      const submitButton = screen.getByTestId('submit-button')
      await userEvent.click(submitButton)

      expect(useCloneEnvironmentMockSpy().mutateAsync).toHaveBeenCalledWith({
        environmentId: mockEnv.id,
        payload: {
          cluster_id: mockClusters[2].id,
          mode: EnvironmentModeEnum.STAGING,
          name: 'test',
        },
      })
    })
  })
})
