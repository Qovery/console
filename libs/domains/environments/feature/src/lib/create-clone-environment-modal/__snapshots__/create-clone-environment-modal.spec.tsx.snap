// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCloneEnvironmentModal should match snapshots 1`] = `
<body>
  <div>
    <div
      class="p-5"
    >
      <h2
        class="h4 max-w-sm truncate text-neutral-400 dark:text-neutral-50"
      >
        Create Environment
      </h2>
      <p
        class="mt-2 text-sm text-neutral-350 dark:text-neutral-50"
      >
        Create a new environment and deploy your applications.
      </p>
      <span
        aria-controls="radix-:rd:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="mt-2 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600"
        data-state="closed"
        type="button"
      >
        Show how it works 
        <i
          aria-hidden="true"
          class="fa-regular fa-circle-question text-xs"
        />
      </span>
      <form
        class="mt-6"
      >
        <div
          class="mb-6"
          data-testid="input-text-wrapper"
        >
          <div
            class="relative"
          >
            <div
              aria-label="input-container"
              class="input     "
            >
              <div
                class=""
              >
                <label
                  class="translate-y-2 text-sm"
                  for="Environment name"
                >
                  Environment name
                </label>
                <input
                  class="input__value "
                  data-testid="input-text"
                  id="Environment name"
                  name="name"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="mb-6"
        >
          <div
            class="input input--select    "
            data-testid="input-select-cluster"
          >
            <label
              class="top-1.5 translate-y-2 text-sm"
              for="Cluster"
            >
              Cluster
            </label>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-4-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class="input-select__control css-13cymwt-control"
              >
                <div
                  class="input-select__value-container css-1fdsijx-ValueContainer"
                >
                  <div
                    class="input-select__placeholder css-1jqq78o-placeholder"
                    id="react-select-4-placeholder"
                  >
                    Select...
                  </div>
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-4-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-readonly="true"
                    class="css-mohuvp-dummyInput-DummyInput"
                    id="Cluster"
                    inputmode="none"
                    role="combobox"
                    tabindex="0"
                    value=""
                  />
                </div>
                <div
                  class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="Cluster"
                type="hidden"
                value=""
              />
            </div>
            <input
              name="Cluster"
              type="hidden"
              value=""
            />
            <div
              class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
            >
              <span
                class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                role="img"
              />
            </div>
          </div>
        </div>
        <div
          class="mb-6"
        >
          <div
            class="input input--select input--has-icon   "
            data-testid="input-select-mode"
          >
            <div
              class="absolute left-0 top-0 flex h-full w-12 items-center justify-center"
              data-testid="selected-icon"
            >
              <span
                class="items-center shrink-0 text-xs border rounded border-neutral-250 text-neutral-350 dark:border-neutral-350 dark:text-neutral-300 flex h-4 w-4 justify-center p-0 font-semibold"
                data-accent-color="neutral"
              >
                D
              </span>
            </div>
            <label
              class="!translate-y-0 !text-xs ml-8"
              for="Type"
            >
              Type
            </label>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-5-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class="input-select__control css-13cymwt-control"
              >
                <div
                  class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
                >
                  <span
                    class="mr-1 text-sm text-neutral-400"
                  >
                    Development
                  </span>
                  <input
                    aria-autocomplete="list"
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-readonly="true"
                    class="css-mohuvp-dummyInput-DummyInput"
                    id="Type"
                    inputmode="none"
                    role="combobox"
                    tabindex="0"
                    value=""
                  />
                </div>
                <div
                  class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="Type"
                type="hidden"
                value="DEVELOPMENT"
              />
            </div>
            <input
              name="Type"
              type="hidden"
              value="DEVELOPMENT"
            />
            <div
              class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
            >
              <span
                class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                role="img"
              />
            </div>
          </div>
        </div>
        <div
          class="mt-6 flex justify-end gap-3"
        >
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-11 px-5 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
            data-testid="cancel-button"
            type="button"
          >
            Cancel
          </button>
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
            data-testid="submit-button"
            disabled=""
            type="submit"
          >
            Create
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
